File: glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package mozilla.telemetry.glean

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Dispatchers as KotlinDispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeoutOrNull
import org.junit.Test
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotSame
import org.junit.Assert.assertSame
import org.junit.Assert.assertTrue
import java.util.concurrent.atomic.AtomicInteger

@Suppress("EXPERIMENTAL_API_USAGE")
class DispatchersTest {

    @Test
    fun `API scope runs off the main thread`() {
        val mainThread = Thread.currentThread()
        var threadCanary = false
        Dispatchers.API.setTestingMode(false)
        Dispatchers.API.setTaskQueueing(false)

        runBlocking {
            Dispatchers.API.launch {
                assertNotSame(mainThread, Thread.currentThread())
                // Use the canary bool to make sure this is getting called before
                // the test completes.
                assertEquals(false, threadCanary)
                threadCanary = true
            }!!.join()
        }

        Dispatchers.API.setTestingMode(true)
        assertEquals(true, threadCanary)
        assertSame(mainThread, Thread.currentThread())
    }

    @Test
    fun `launch correctly adds tasks to queue if queueTasks is true`() {
        var threadCanary = 0

        Dispatchers.API.setTestingMode(true)
        Dispatchers.API.setTaskQueueing(true)

        // Add 3 tasks to queue each one setting threadCanary to true to indicate if any task has ran
        repeat(3) {
            Dispatchers.API.launch {
                threadCanary += 1
            }
        }

        assertEquals("Task queue contains the correct number of tasks",
            3, Dispatchers.API.taskQueue.size)
        assertEquals("Tasks have not run while in queue", 0, threadCanary)

        // Now trigger execution to ensure the tasks fired
        runBlocking {
            Dispatchers.API.flushQueuedInitialTasks()
        }

        assertEquals("Tasks have executed", 3, threadCanary)
        assertEquals("Task queue is cleared", 0, Dispatchers.API.taskQueue.size)
    }

    @Test
    fun `queued tasks are flushed off the main thread`() {
        val mainThread = Thread.currentThread()
        val threadCanary = AtomicInteger()

        // By setting testing mode to false, we make sure that things
        // are executed asynchronously.
        Dispatchers.API.setTestingMode(false)
        Dispatchers.API.setTaskQueueing(true)

        // Add 3 tasks to queue each one setting threadCanary to true
        // to indicate if any task has ran.
        repeat(3) {
            Dispatchers.API.launch {
                assertNotSame(
                    "Pre-init tasks must be flushed off the main thread",
                    mainThread,
                    Thread.currentThread()
                )
                threadCanary.incrementAndGet()
            }
        }

        assertEquals("Task queue contains the correct number of tasks",
            3, Dispatchers.API.taskQueue.size)
        assertEquals("Tasks have not run while in queue", 0, threadCanary.get())

        // Now trigger execution to ensure the tasks fired
        GlobalScope.launch {
            Dispatchers.API.flushQueuedInitialTasks()
        }

        // Wait for the flushed tasks to be executed.
        runBlocking {
            withTimeoutOrNull(2000) {
                while (isActive && (threadCanary.get() != 3 || Dispatchers.API.taskQueue.size > 0)) {
                    delay(1)
                }
            } ?: assertTrue("Timed out waiting for tasks to execute", false)
        }

        assertEquals("All tasks ran to completion", 3, threadCanary.get())
        assertEquals("Task queue is cleared", 0, Dispatchers.API.taskQueue.size)
    }

    @Test
    fun `queued tasks are executed in the order they are received`() {
        val orderedList = mutableListOf<Int>()
        val jobs = mutableListOf<Job>()

        Dispatchers.API.setTestingMode(false)
        Dispatchers.API.setTaskQueueing(true)

        val coroutineScope = CoroutineScope(KotlinDispatchers.Default)

        // This coroutine will monitor the taskQueue.count() to toggle the flushing of the queued
        // items when the queue is half full (50 elements).  This should give us 50 items in the
        // queue and then 50 items that are launched after the queue is flushed.
        val flushJob = coroutineScope.launch {
            while (Dispatchers.API.taskQueue.count() < 50) { Thread.yield() }
            Dispatchers.API.flushQueuedInitialTasks()
        }

        // This coroutine will add elements to the orderedList.  This will continue to
        // add elements to the queue until there are at least 50 elements in the queue. At that
        // point, the coroutine above will flush and disable the queuing and this coroutine will
        // continue launching tasks directly.
        // Note: we need a counter to make sure that all the jobs are added to the `jobs` array.
        // we can't simply check the array size as it might change while we're checking.
        val counter = AtomicInteger()
        val listJob = coroutineScope.launch(KotlinDispatchers.Default) {
            (0..99).forEach { num ->
                Dispatchers.API.launch {
                    orderedList.add(num)
                    counter.incrementAndGet()
                }?.let {
                    jobs.add(it)
                }
            }
        }

        // Wait for the numbers to be added to the list by waiting for the tasks to join.
        runBlocking {
            // Ensure that all the required jobs have been added to the list.
            withTimeoutOrNull(2000) {
                while (isActive && counter.get() < 100) {
                    delay(1)
                }
            } ?: assertEquals("Timed out waiting for tasks to execute", 100, counter.get())

            // Wait for them to execute.
            flushJob.join()
            listJob.join()
            jobs.joinAll()
        }

        // Ensure elements match in the correct order
        (0..99).forEach { num ->
            assertTrue(
                "Index [$num] is less than size of list [${orderedList.size}]",
                num < orderedList.size
            )
            assertEquals("This list is out of order $orderedList", num, orderedList[num])
        }
    }

    @Test
    fun `dispatched tasks throwing exceptions are correctly handled`() {
        val mainThread = Thread.currentThread()
        val threadCanary = AtomicInteger()

        // By setting testing mode to false, we make sure that things
        // are executed asynchronously.
        Dispatchers.API.setTestingMode(false)
        Dispatchers.API.setTaskQueueing(false)

        // Dispatch an initial tasks that throws an exception.
        Dispatchers.API.launch {
            assertNotSame(
                "Tasks must be executed off the main thread",
                mainThread,
                Thread.currentThread()
            )
            @Suppress("TooGenericExceptionThrown")
            throw Exception("Test exception for DispatchersTest")
        }

        // Add 3 tasks to queue each one increments threadCanary
        // to indicate if any task has ran.
        repeat(3) {
            Dispatchers.API.launch {
                assertNotSame(
                    "Tasks must be executed off the main thread",
                    mainThread,
                    Thread.currentThread()
                )
                threadCanary.incrementAndGet()
            }
        }

        // Wait for the flushed tasks to be executed.
        runBlocking {
            withTimeoutOrNull(2000) {
                while (isActive && (threadCanary.get() != 3)) {
                    delay(1)
                }
            } ?: assertTrue("Timed out waiting for tasks to execute", false)
        }

        assertEquals("All the dispatched actions should execute", 3, threadCanary.get())
    }
}

File: glean-core/ios/GleanTests/DispatchersTest.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

@testable import Glean
import XCTest

class DispatchersTest: XCTestCase {
    func testTaskQueuing() {
        var threadCanary = 0

        Dispatchers.shared.setTestingMode(enabled: true)
        Dispatchers.shared.setTaskQueueing(enabled: true)

        // Add 3 tasks to the queue, each one incrementing threadCanary to indicate the task has executed
        for _ in 0 ..< 3 {
            Dispatchers.shared.launchAPI {
                threadCanary += 1
            }
        }

        XCTAssertEqual(
            Dispatchers.shared.preInitOperations.count,
            3,
            "Task queue contains the correct number of tasks"
        )
        XCTAssertEqual(
            threadCanary,
            0,
            "Tasks have not executed while in queue"
        )

        // Now trigger the queue to fire the tasks
        Dispatchers.shared.flushQueuedInitialTasks()

        XCTAssertEqual(
            threadCanary,
            3,
            "Tasks have executed"
        )
        XCTAssertEqual(
            Dispatchers.shared.preInitOperations.count,
            0,
            "Task queue has cleared"
        )
    }

    func testQueuedTasksAreExecutedInOrder() {
        // Create a test operation queue to run our operations asynchronously
        let testQueue = OperationQueue()
        testQueue.name = "Glean test queue"
        // Set to 2 to allow both of our tasks to run concurrently
        testQueue.maxConcurrentOperationCount = 2

        var orderedList = [Int]()
        var flushTasks = AtomicBoolean(false)

        Dispatchers.shared.setTestingMode(enabled: false)
        Dispatchers.shared.setTaskQueueing(enabled: true)

        // This background task will monitor the size of the cached initial
        // operations and attempt to flush it when it reaches 50 elements.
        // This should give us 50 items in the queued items and 50 that are
        // executed in the background normally.
        let op1 = BlockOperation {
            while !flushTasks.value { sleep(1) }
            Dispatchers.shared.flushQueuedInitialTasks()
        }
        testQueue.addOperation(op1)

        // This background task will add elements to the orderedList.  This will continue to
        // add elements to the queue until there are at least 50 elements in the queue. At that
        // point, the background task above will flush and disable the queuing and this task will
        // continue launching tasks directly.
        var counter = 0
        let op2 = BlockOperation {
            for num in 0 ... 99 {
                if num == 50 {
                    flushTasks.value = true
                }
                Dispatchers.shared.launchAPI {
                    orderedList.append(num)
                    counter += 1
                }
            }
        }
        testQueue.addOperation(op2)

        // Wait for the numbers to be added to the list by waiting for the operations above to complete
        op1.waitUntilFinished()
        op2.waitUntilFinished()

        // Wait for all of the elements to be added to the list before we check the ordering
        while counter < 100 { sleep(1) }

        // Make sure the elements match in the correct order
        for num in 0 ... 99 {
            XCTAssertEqual(
                num,
                orderedList[num],
                "This list is out of order, \(num) != \(orderedList[num])"
            )
        }
    }

    func testCancelBackgroundTasksClearsQueuedItems() {
        // Set testing mode to false to allow for background execution
        Dispatchers.shared.setTestingMode(enabled: false)

        // Set task queueing to true to ensure that we queue tasks when we launch them
        Dispatchers.shared.setTaskQueueing(enabled: true)

        // Add a task to the pre-init queue
        Dispatchers.shared.launchAPI {
            print("A queued task")
        }

        // Assert the task was queued
        XCTAssertEqual(Dispatchers.shared.preInitOperations.count, 1, "Task must be queued")

        // Now cancel the tasks, which will also clear the queued initial tasks
        Dispatchers.shared.cancelBackgroundTasks()

        // Assert the task was removed from the queue
        XCTAssertEqual(Dispatchers.shared.preInitOperations.count, 0, "Task must be removed")
    }

    func testCancelBackgroundTasksSerialQueue() {
        // Set testing mode to false to allow for background execution
        Dispatchers.shared.setTestingMode(enabled: false)

        // Set up our test conditions for normal execution by setting queuing to false
        Dispatchers.shared.setTaskQueueing(enabled: false)

        // Create a counter to use to determine if the task is actually cancelled
        var counter = 0

        // Create a task to add to the Dispatchers serial queue that looks at the
        // `isCancelled` property so that it can be cancelled
        let serialOperation = BlockOperation()
        serialOperation.addExecutionBlock {
            while !serialOperation.isCancelled {
                counter += 1
            }
        }
        Dispatchers.shared.serialOperationQueue.addOperation(serialOperation)

        // Let the task run for 1 second in order to it time to accumulate to the counter
        sleep(1)

        // Check that the counter has incremented
        XCTAssertTrue(counter > 0, "Serial task must execute")

        // Now cancel the background task
        Dispatchers.shared.cancelBackgroundTasks()

        // Wait for the task to be cancelled/finished, the OperationQueue will set
        // `isFinished` for cancelled tasks
        Dispatchers.shared.serialOperationQueue.waitUntilAllOperationsAreFinished()

        // Grab the current counter value. This shouldn't change after the task was cancelled
        let testValue = counter

        // Wait for one second to ensure task is truly cancelled. This gives the background task
        // time to continue to run and accumulate to the counter in the case that it wasn't
        // actually cancelled, that way we can detect the failure to cancel by comparing to the
        // snapshot of the counter value we just captured.
        sleep(1)

        // Make sure counters haven't changed
        XCTAssertEqual(counter, testValue, "Serial task must be cancelled")
    }

    func testCancelBackgroundTasksClearsConcurrentQueue() {
        // Set testing mode to false to allow for background execution
        Dispatchers.shared.setTestingMode(enabled: false)

        // Set up our test conditions for normal execution by setting queuing to false
        Dispatchers.shared.setTaskQueueing(enabled: false)

        // Create a counter to use to determine if the task is actually cancelled
        var counter = 0

        // Create a task to add to the concurrently executed queue that looks at the
        // `isCancelled` property so that it can be cancelled
        let concurrentOperation = BlockOperation()
        concurrentOperation.addExecutionBlock {
            while !concurrentOperation.isCancelled {
                counter += 1
            }
        }
        Dispatchers.shared.launchConcurrent(operation: concurrentOperation)

        // Let the task run for 1 second in order to give it time accumulate to the counter
        sleep(1)

        // Check that the counter has incremented
        XCTAssertTrue(counter > 0, "Concurrent task must execute")

        // Now cancel the background tasks
        Dispatchers.shared.cancelBackgroundTasks()

        // Wait for the task to be cancelled/finished, the OperationQueue will set
        // `isFinished` for cancelled tasks
        Dispatchers.shared.concurrentOperationsQueue.waitUntilAllOperationsAreFinished()

        // Grab the current counter value. This shouldn't change after the task is cancelled
        let asyncTest = counter

        // Wait for one second to ensure task is truly cancelled. This gives the background task
        // time to continue to run and accumulate to the counter in the case that it wasn't
        // actually cancelled, that way we can detect the failure to cancel by comparing to the
        // snapshot of the counter value we just captured.
        sleep(1)

        // Make sure counter hasn't changed
        XCTAssertEqual(counter, asyncTest, "Concurrent task must be cancelled")
    }

    func testOverflowingTheTaskQueueRecordsTelemetry() {
        Glean.shared.resetGlean(clearStores: true)
        Dispatchers.shared.preInitOperations.removeAll()
        Dispatchers.shared.setTestingMode(enabled: true)
        Dispatchers.shared.setTaskQueueing(enabled: true)

        for _ in 0 ..< (Dispatchers.Constants.maxQueueSize + 10) {
            Dispatchers.shared.launchAPI {
                // Do nothing
            }
        }

        XCTAssertEqual(
            Int(Dispatchers.Constants.maxQueueSize),
            Dispatchers.shared.preInitOperations.count,
            "Task queue must contain the maximum number of tasks"
        )
        XCTAssertEqual(
            Dispatchers.Constants.maxQueueSize + 10,
            Dispatchers.shared.preInitTaskCount,
            "preInitTaskCount is correct"
        )

        // Resetting Glean here causes the tasks to be flushed and the
        // queueing turned off, this should cause the error metric to
        // be recorded so we can check it to ensure the value matches
        // the expected value.
        Glean.shared.resetGlean(clearStores: false)

        XCTAssertEqual(
            Dispatchers.Constants.maxQueueSize + 10,
            try GleanMetrics.GleanError.preinitTasksOverflow.testGetValue(),
            "preInitTaskCount is correct"
        )

        Glean.shared.resetGlean(clearStores: true)
    }
}
